uniffi_core/src/ffi/rustfuture.rs

```diff
@@ -0,0 +1,642 @@
+                .into_unit_ptr();
+        let foreign_waker = unsafe { RustFutureForeignWaker::from_unit_ptr(foreign_waker_ptr) };
+
+        // Clone to increase the strong count, so that we can see if it has not been dropped by `wake_by_ref` later.
+        let _ = unsafe { RustFutureForeignRawWaker::clone_waker(foreign_waker_ptr) };
+        assert_eq!(Arc::strong_count(&foreign_waker), 2);
+
+        // Let's call the waker by reference.
+        unsafe { RustFutureForeignRawWaker::wake_by_ref(foreign_waker_ptr) };
+
+        // Has it been called?
+        let foreign_waker_environment = unsafe { Box::from_raw(foreign_waker_environment_ptr) };
+        assert!(foreign_waker_environment.take());
+
+        // Has the waker not been dropped?
+        assert_eq!(Arc::strong_count(&foreign_waker), 2);
+
+        // Dropping manually to avoid data leak.
+        unsafe { RustFutureForeignRawWaker::drop_waker(foreign_waker_ptr) };
+    }
+}
+
+const READY: bool = true;
+const PENDING: bool = false;
+
+/// Poll a [`RustFuture`]. If the `RustFuture` is ready, the function returns
+/// `true` and puts the result inside `polled_result`, otherwise it returns
+/// `false` and _doesn't modify_ the value inside `polled_result`. A third
+/// case exists: if the `RustFuture` is throwing an error, the function returns
+/// `true` but doesn't modify `polled_result` either, however the value
+/// of  `call_status` is changed appropriately. It is summarized inside the
+/// following table:
+///
+/// | `RustFuture`'s state | `polled_result`         | `call_status.code` | returned value |
+/// |----------------------|-------------------------|--------------------|----------------|
+/// | `Ready(Ok(T))`       | is mapped to `T::lower` | `CALL_SUCCESS`     | `true`         |
+/// | `Ready(Err(E))`      | is not modified         | `CALL_ERROR`       | `true`         |
+/// | `Pending`            | is not modified         | `CALL_SUCCESS`     | `false`        |
+///
+/// Please see the module documentation to learn more.
+///
+/// # Panics
+///
+/// The function panics if `future` or `waker` is a NULL pointer.
+pub fn uniffi_rustfuture_poll<T, E>(
+    future: Option<&mut RustFuture<T, E>>,
+    waker: Option<RustFutureForeignWakerFunction>,
+    waker_environment: *const RustFutureForeignWakerEnvironment,
+    polled_result: &mut T::FfiType,
+    call_status: &mut RustCallStatus,
+) -> bool
+where
+    T: FfiReturn,
+    E: FfiError + FfiConverter<RustType = E, FfiType = RustBuffer>,
+{
+    // If polling the future panics, an error will be recorded in call_status and the future will
+    // be dropped, so there is no potential for observing any inconsistent state in it.
+    let mut future = panic::AssertUnwindSafe(future.expect("`future` is a null pointer"));
+    let waker = waker.expect("`waker` is a null pointer");
+
+    match call_with_result(call_status, move || {
```

Comment:
Is the way we define `call_with_result` vs `call_with_output` is why you made `RustFuture` only work with `Result<>` types?  If so, maybe I can refactor some things to simplify this for you.

Reply:
We need `RustFuture` to output a `Result` to handle errors with the same code. Otherwise the macros and the code become more complex. `call_with_output` redirect to `call_with_result` anyway, we apply the same transformation here.